core/src/Streamly/Internal/Data/Array.hs

```diff
@@ -175,50 +171,38 @@ fromList xs = unsafePerformIO $ fromStreamD $ D.fromList xs
 
 {-# INLINE length #-}
 length :: Array a -> Int
-length = sizeofArray
+length = arrLen
 
 {-# INLINE_NORMAL read #-}
-read :: Monad m => Unfold m (Array a) a
-read = Unfold step inject
-  where
-    inject arr = return (arr, 0)
-    step (arr, i)
-        | i == length arr = return D.Stop
-    step (arr, I# i) =
-        return $
-        case Exts.indexArray# (array# arr) i of
-            (# x #) -> D.Yield x (arr, I# i + 1)
+read :: MonadIO m => Unfold m (Array a) a
+read = Unfold.lmap unsafeThaw MArray.read
 
 -------------------------------------------------------------------------------
 -- Elimination - to streams
 -------------------------------------------------------------------------------
 
+{-# INLINE_NORMAL toList #-}
+toList :: Array a -> [a]
+toList arr = loop 0
+
+    where
+
+    len = length arr
+    loop i | i == len = []
+    loop i = getIndexUnsafe arr i : loop (i + 1)
+
 {-# INLINE_NORMAL toStreamD #-}
-toStreamD :: Monad m => Array a -> D.Stream m a
-toStreamD arr = D.Stream step 0
-  where
-    {-# INLINE_LATE step #-}
-    step _ i
-        | i == length arr = return D.Stop
-    step _ (I# i) =
-        return $
-        case Exts.indexArray# (array# arr) i of
-            (# x #) -> D.Yield x (I# i + 1)
+toStreamD :: MonadIO m => Array a -> D.Stream m a
+toStreamD = MArray.toStreamD . unsafeThaw
 
 {-# INLINE_NORMAL toStreamDRev #-}
 toStreamDRev :: Monad m => Array a -> D.Stream m a
-toStreamDRev arr = D.Stream step (length arr - 1)
-  where
-    {-# INLINE_LATE step #-}
-    step _ i
-        | i < 0 = return D.Stop
-    step _ (I# i) =
-        return $
-        case Exts.indexArray# (array# arr) i of
-            (# x #) -> D.Yield x (I# i - 1)
+toStreamDRev arr@Array{..} =
+    D.map (getIndexUnsafe arr)
+        $ D.enumerateFromThenToIntegral (arrLen - 1) (arrLen - 2) 0
```

Comment:
We could implement it like `MArray.toStreamDRev . unsafeThaw`

Reply:
Yes, We can move a lot of functions to the `Mut`. But I'm holding this off in this change.